### PR TITLE
Add optional colors for kubectl printing...

### PR DESCRIPTION
**Pull request by brendandburns**
@kubernetes/sig-cli-pr-reviews 

This PR adds an initial path for optional color support.

Currently:

```
kubectl get pods --color
```

Will print colorized output.

Eventually, I will add an environment variable as well.  (and colors for other things)

### CODE REVIEW

**Comment by brendandburns** [comment]
colors

**Comment by smarterclayton** [comment]
Why color the name column instead of the status column?  The users eye will need to go the status column anyway on a failure, and so this actually makes it harder to spot invalid items in that case (makes the user do more work, because the color will pull the eye).  Likewise, for normal operation the color against a black background makes the name harder to read in the "normal" case, which makes the user do more work to scan a list.

**Comment by smarterclayton** [comment]
Generally you want color to pull the eye to the spot that is abnormal, not to anchor it on the things that are "normal".  I might suggest we want the entire row to be bold on error, and normal weight on normal status.

Finally, all this is moving to the server side, so doing this now is putting extra weight on clients now that complicates that design (or is at minimum an extra requirement).  How is this going to work for third party resources?  I'd like to keep this really minimal for right now, and not complicate those other efforts that are actually required in the short term.

**Comment by brendandburns** [comment]
@smarterclayton we're moving human readable printing server side? That can't be right...

Feedback on the colors and coloring the name vs. the status makes sense to me, though I think there's a degree to wanting a measure of the pod "health" in general.

e.g. number of restarts, number of containers running, etc. all contribute to a pods "healthiness"

I worry that we'll end up with lots of different colors if we don't focus on coloring the entire row a single shade...

(I'd be open to coloring the entire row, or restricting coloring only to "abnormal" rows)

**Comment by kargakis** [comment]
Is this going to work consistently across different terminals? Is it something that has been discussed in @kubernetes/sig-cli-misc ?

File: pkg/kubectl/resource_printer.go
Comment (juanvallejo) [inline pkg/kubectl/resource_printer.go]: @brendandburns 
> I worry that we'll end up with lots of different colors if we don't focus on coloring the entire row a single shade...

Based on this comment, and @smarterclayton's feedback, perhaps we could keep red for rows where the pod status is failing, and no color / how it is currently for any other case. WDYT?

Comment (brendandburns) [inline pkg/kubectl/resource_printer.go]: Yeah, that works for me. I'd love to have yellow for "warning" (pending, some restarts recently, etc) and "red" for "not working"

I'll update the pr.

**Comment by brendandburns** [comment]
@kargakis this should work across all Xterm capable terminals. It is off by default as well (and I think it will be off by default for a long time) so I think we don't have to have perfect fidelity everywhere, since it will only affect people who opt into it.

**Comment by brendandburns** [comment]
Also, @smarterclayton considering the server side bits, the colors actually come as a markup language, so I think that will actually make it easier for the server side, since the server can return markup colors and the client can figure out how to render them.

**Comment by soltysh** [comment]
My $0.02, why coloring would ever belong to server side? I can understand the desire to have a consistent output across client apps, but coloring should belong to client app, imho. The server side printing is a whole different story and I do have a mixed feelings about it, too. 

👍 for having color emphasize the abnormal state and none for ok.

**Comment by brendandburns** [comment]
@soltysh I agree entirely, and honestly this is why I have significant concerns about server-side printing.

However, if we move table-based printing to the server, the information needed to do the color-based decision making will be lost and stored on the server, not in the client, so the server would be the only place to make that decision.

**Comment by brendandburns** [comment]
Friendly ping on this.

Thanks
--brendan

**Comment by fabianofranz** [comment]
I like the idea of making it opt-in at least until it stabilizes.

Any strong reason for choosing that particular library? When I R&D about colors a few months ago I found https://github.com/fatih/color, it supports Windows `cmd.exe` and can also make the existing `fmt` printing functions colorizable (although I didn't check how that was implemented). Not strong about one or another though.

**Comment by fabianofranz** [comment]
@k8s-bot test this

**Comment by soltysh** [comment]
> However, if we move table-based printing to the server, the information needed to do the color-based decision making will be lost and stored on the server, not in the client, so the server would be the only place to make that decision.

That will only be true for the extended printing options, that @smarterclayton is proposing, right?

File: pkg/kubectl/resource_printer.go
Comment (soltysh) [inline pkg/kubectl/resource_printer.go]: Still haven't seen this one addressed. Additionally if I add watch flag I'm getting:
```
$ kubectl get pods --color -w
NAME          READY     STATUS    RESTARTS   AGE
<fg 2>hello-5b209<reset>   1/1       Running   0          18s
```

and that's definitely not what I'd expect.

**Review by soltysh — CHANGES_REQUESTED** [review]

**Comment by brendandburns** [comment]
@fabianofranz I actually started this PR using https://github.com/fatih/color.

The trouble that I ran into was that the colorization characters broke the golang TabWriter class, see this question: http://stackoverflow.com/questions/35398497/how-do-i-get-colors-to-work-with-golang-tabwriter for someone who ran into the same problem.

Because this library uses markup, and you can train `TabWriter` to ignore HTML characters this library actually works and also does the right things with indents.

@soltysh will work on getting the comments/problems addressed.

Thanks!

**Comment by brendandburns** [comment]
fwiw, it turns out we already colorize:

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/clusterinfo.go#L126

**Comment by smarterclayton** [comment]
I'd like this to be done under the context of moving to the server side,
rather than be add before it lands and complicate it.  I think color will
help some, but I don't want to add more features to printing until the
proposals are approved for server side, and no one is broken without it.
Hopefully this has enough context that you can describe in the proposal
what usecases are necessary (at least at a first glance, we want to flag
cells with informational bits like "alert" or "warn").

On Mar 13, 2017, at 12:49 AM, Brendan Burns <notifications@github.com>
wrote:

fwiw, it turns out we already colorize:

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/clusterinfo.go#L126

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/40963#issuecomment-286014973>,
or mute the thread

**Comment by smarterclayton** [comment]
As the carrot, the backend glue is all prototyped, so working together
could get it into place fairly quickly.

On Mar 13, 2017, at 12:49 AM, Brendan Burns <notifications@github.com>
wrote:

fwiw, it turns out we already colorize:

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/clusterinfo.go#L126

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/40963#issuecomment-286014973>,
or mute the thread

**Comment by brendandburns** [comment]
@smarterclayton where is the proposal for server-side printing?

I'm happy to pile on there.

**Comment by soltysh** [comment]
@brendandburns proposal: https://github.com/kubernetes/community/pull/363, prototype: https://github.com/kubernetes/kubernetes/pull/40848

File: pkg/kubectl/cmd/util/factory_object_mapping.go
Comment (bgrant0607) [inline pkg/kubectl/cmd/util/factory_object_mapping.go]: Does this detect whether a terminal is attached or not, as well?

File: pkg/kubectl/cmd/get.go
Comment (bgrant0607) [inline pkg/kubectl/cmd/get.go]: AFAICT, this doesn't have a license. We can't merge this.

We need a verify script to look for this type of thing. It's repeatedly caused us grief.

Comment (bgrant0607) [inline pkg/kubectl/cmd/get.go]: Filed #44505

File: pkg/kubectl/resource_printer.go
Comment (bgrant0607) [inline pkg/kubectl/resource_printer.go]: I understand the benefit, but we need to reconcile this with https://github.com/kubernetes/community/pull/363

**Comment by bgrant0607** [comment]
BTW, I also want to nuke the kubeconfig preferences section, which contains only the colors preference currently. #10693

**Pull request closed without merging**
